core/math/vertex_cache_optimizer.h

```diff
@@ -0,0 +1,120 @@
+/* SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.                 */
+/**************************************************************************/
+
+#ifndef VERTEX_CACHE_OPTIMIZER_H
+#define VERTEX_CACHE_OPTIMIZER_H
+
+// This class is derived from
+// https://www.martin.st/thesis/
+// Based on Tom Forsyth's vertex cache optimizer
+
+/*
+  Copyright (C) 2008 Martin Storsjo
+  This software is provided 'as-is', without any express or implied
+  warranty.  In no event will the authors be held liable for any damages
+  arising from the use of this software.
+  Permission is granted to anyone to use this software for any purpose,
+  including commercial applications, and to alter it and redistribute it
+  memfreely, subject to the following restrictions:
+  1. The origin of this software must not be misrepresented; you must not
+	 claim that you wrote the original software. If you use this software
+	 in a product, an acknowledgment in the product documentation would be
+	 appreciated but is not required.
+  2. Altered source versions must be plainly marked as such, and must not be
+	 misrepresented as being the original software.
+  3. This notice may not be removed or altered from any source distribution.
+*/
+
+#include "core/local_vector.h"
+#include "core/math/math_defs.h"
+
+#include <stdint.h>
+
+class VertexCacheOptimizer {
+	typedef uint32_t VERTEX_INDEX_TYPE;
+
+	// The size of these data types affect the memory usage.
+	typedef uint16_t SCORE_TYPE;
+	typedef uint8_t ADJACENCY_TYPE;
+	typedef int8_t CACHE_POS_TYPE;
+	typedef int32_t TRIANGLE_INDEX_TYPE;
+	typedef int32_t ARRAY_INDEX_TYPE;
+
+	struct Constants {
+		// The size of the precalculated tables.
+		static const int CACHE_SCORE_TABLE_SIZE = 32;
+		static const int VALENCE_SCORE_TABLE_SIZE = 32;
+		static const int MAX_ADJACENCY = UINT8_MAX;
+		static const int SCORE_SCALING = 7281;
+
+		// Score function constants.
+		static constexpr float CACHE_DECAY_POWER = 1.5;
+		static constexpr float LAST_TRI_SCORE = 0.75;
+		static constexpr float VALENCE_BOOST_SCALE = 2.0;
+		static constexpr float VALENCE_BOOST_POWER = 0.5;
+
+		// Set these to adjust the performance and result quality.
+		static const int VERTEX_CACHE_SIZE = 24;
+		static const int CACHE_FUNCTION_LENGTH = 32;
+
+		static_assert(CACHE_SCORE_TABLE_SIZE >= VERTEX_CACHE_SIZE, "Vertex score table too small");
+	};
+
+	// Precalculated tables.
+	SCORE_TYPE _cache_position_score[Constants::CACHE_SCORE_TABLE_SIZE];
+	SCORE_TYPE _valence_score[Constants::VALENCE_SCORE_TABLE_SIZE];
+
+	int is_added(const uint8_t *p_triangle_added, int p_x) const {
+		return p_triangle_added[(p_x) >> 3] & (1 << (p_x & 7));
+	}
+
+	void set_added(uint8_t *p_triangle_added, int p_x) const {
+		p_triangle_added[(p_x) >> 3] |= (1 << (p_x & 7));
+	}
+
+	// Precalculate the tables.
+	void init();
+
+	// Calculate the score for a vertex.
+	SCORE_TYPE find_vertex_score(int numActiveTris, int cachePosition);
```

Comment:
```suggestion
	SCORE_TYPE find_vertex_score(int p_num_active_tris, int p_cache_position);
```

Reply:
Oh yep I changed this in the cpp but QTCreator rename isn't always good at changing cpp + header with the refactor function. :grinning: 
Will change.